### ISSUE TITLE
Array compute method fails with "TypeError: unhashable type: 'list'" with resource kwargs

### ISSUE THREAD

**Issue by sjperkins** [dask/distributed#2027]
python 2.7.12
distributed 1.21.8
dask 0.17.5

The following:

```python
from dask.core import flatten
import dask.array as da
from distributed import Client

c = Client(resources={'a':2})

A = da.ones((100,100), chunks=(10,10))
akeys = tuple(flatten(A.__dask_keys__()))
A.compute(resources={akeys:{'a':2}})
```

produces the following stacktrace

```python-stacktrace
distributed.protocol.core - CRITICAL - Failed to deserialize
Traceback (most recent call last):
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 96, in loads
    msg = loads_msgpack(small_header, small_payload)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 173, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "msgpack/_unpacker.pyx", line 200, in msgpack._unpacker.unpackb
TypeError: unhashable type: 'list'
distributed.scheduler - ERROR - unhashable type: 'list'
Traceback (most recent call last):
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/scheduler.py", line 1929, in handle_client
    msgs = yield comm.read()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/concurrent.py", line 260, in result
    raise_exc_info(self._exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 1107, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/tcp.py", line 203, in read
    msg = yield from_frames(frames, deserialize=self.deserialize)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/concurrent.py", line 260, in result
    raise_exc_info(self._exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 315, in wrapper
    yielded = next(result)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/utils.py", line 75, in from_frames
    res = _from_frames()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/utils.py", line 61, in _from_frames
    return protocol.loads(frames, deserialize=deserialize)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 96, in loads
    msg = loads_msgpack(small_header, small_payload)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 173, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "msgpack/_unpacker.pyx", line 200, in msgpack._unpacker.unpackb
TypeError: unhashable type: 'list'
distributed.utils - ERROR - unhashable type: 'list'
Traceback (most recent call last):
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/utils.py", line 622, in log_errors
    yield
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/client.py", line 921, in _handle_report
    six.reraise(*clean_exception(**msg))
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/concurrent.py", line 260, in result
    raise_exc_info(self._exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 1107, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/tcp.py", line 203, in read
    msg = yield from_frames(frames, deserialize=self.deserialize)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/concurrent.py", line 260, in result
    raise_exc_info(self._exc_info)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/tornado/gen.py", line 315, in wrapper
    yielded = next(result)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/utils.py", line 75, in from_frames
    res = _from_frames()
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/comm/utils.py", line 61, in _from_frames
    return protocol.loads(frames, deserialize=deserialize)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 96, in loads
    msg = loads_msgpack(small_header, small_payload)
  File "/home/sperkins/venv/tricolour/local/lib/python2.7/site-packages/distributed/protocol/core.py", line 173, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "msgpack/_unpacker.pyx", line 200, in msgpack._unpacker.unpackb
TypeError: unhashable type: 'list'
```

**Comment by sjperkins** [comment]
This is fixable by using `client._expand_resources` in `client.get`.

**Comment by mrocklin** [comment]
Interestingly I get the following:

```python
In [1]: from dask.core import flatten
   ...: import dask.array as da
   ...: from distributed import Client
   ...: 
   ...: c = Client(resources={'a':2})
   ...: 
   ...: A = da.ones((100,100), chunks=(10,10))
   ...: akeys = tuple(flatten(A.__dask_keys__()))
   ...: A.compute(resources={akeys:{'a':2}})
   ...: 

Out[1]: 
array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.]])
```

**Comment by martindurant** [comment]
I also see the `TypeError: unhashable type: 'list'` with dask and distributed at master

**Comment by sjperkins** [comment]
Fixed by #2131

**Comment by chrisroat** [comment]
What was the resolution here supposed to be?

I'm seeing this error with on MacOS 10.15.2 with:
```
Click==7.0
cloudpickle==1.2.2
dask==2.9.2
distributed==2.9.3
HeapDict==1.0.1
msgpack==0.6.2
numpy==1.18.1
psutil==5.6.7
PyYAML==5.3
sortedcontainers==2.1.0
tblib==1.6.0
toolz==0.10.0
tornado==6.0.3
zict==1.0.0
```

I tried using the flatten workaround, but forgot the tuple -- the code runs, but it ignores the resources.  I was using `dask.compute` and wasn't sure if that was the problem.  So I actually tested a whole bunch of things to see what worked, and thought it might be helpful.

Below is the code and the results of running every combination of things.  My observations

1. If the unhashable issue had been resolved, it may have been undone
2. Using `flatten(...)` instead of `tuple(flatten(...))` is proceeding silently and not as I expected.  Is throwing an error here an option, or is the result of `flatten(...)` sometimes a correct key?
3. There was no difference between `dask.compute` and `arr.compute`, which was my own red-herring.

```
import dask
from dask.core import flatten
import dask.array as da
from distributed import Client

c = Client()
#c = Client(resources={'a': 2})

A = da.ones((100,100), chunks=(10,10))

#akeys = A.__dask_keys__()
akeys = flatten(A.__dask_keys__())
#akeys = tuple(flatten(A.__dask_keys__()))

# A.compute(resources={akeys:{'a':2}})
dask.compute(A, resources={akeys:{'a':2}})

# client          compute_on      flatten_method    output            notes
# ------          ------------    --------------    ------            -----
# no resources    array           tuple(flatten())  halts             correct
# resources       array           tuple(flatten())  computes          correct
# no resources    array           flatten()         computes          bad syntax leads to wrong behavior
# resources       array           flatten()         computes          computes, but is incorrect syntax
# no resources    array           None              unhashable error  needs bugfix?
# resources       array           None              unhashable error  needs bugfix?

# no resources    dask.compute    tuple(flatten())  halts             correct
# resources       dask.compute    tuple(flatten())  computes          correct
# no resources    dask.compute    flatten()         computes          bad syntax leads to wrong behavior
# resources       dask.compute    flatten()         computes          computes, but is incorrect syntax
# no resources    dask.compute    None              unhashable error  needs bugfix?
# resources       dask.compute    None              unhashable error  needs bugfix?
```
